File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_go-stdlib/play_quests.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"
)

// playQuestCreateRequest binds the payload for a new play-campaign quest.
type playQuestCreateRequest struct {
	QuestID   string   `json:"quest_id"`
	Title     string   `json:"title"`
	DependsOn []string `json:"depends_on"`
}

// questRewards is the configured reward payload embedded in a quest record.
type questRewards struct {
	XP    int            `json:"xp"`
	Items map[string]int `json:"items"`
}

// playQuestResponse is the shape returned when a quest is created, updated,
// or listed. It always includes a non-null depends_on array. Rewards are
// omitted from the JSON when they have not been configured.
type playQuestResponse struct {
	QuestID   string        `json:"quest_id"`
	Title     string        `json:"title"`
	DependsOn []string      `json:"depends_on"`
	State     string        `json:"state"`
	Rewards   *questRewards `json:"rewards,omitempty"`
}

// playQuestListResponse is the shape returned by the quest listing endpoint.
type playQuestListResponse struct {
	Quests []playQuestResponse `json:"quests"`
}

// playQuestStateRequest binds the payload for a quest state transition.
type playQuestStateRequest struct {
	State string `json:"state"`
}

// questRewardsRequest binds the payload for configuring quest rewards.
type questRewardsRequest struct {
	XP    int            `json:"xp"`
	Items map[string]int `json:"items"`
}

// questAwardResponse is the shape returned after awarding quest rewards.
type questAwardResponse struct {
	QuestID string         `json:"quest_id"`
	Awarded bool           `json:"awarded"`
	XP      int            `json:"xp"`
	Items   map[string]int `json:"items"`
}

// characterQuestRewardsResponse is the shape returned when reading a
// character's cumulative quest reward grants.
type characterQuestRewardsResponse struct {
	CharacterID string         `json:"character_id"`
	XP          int            `json:"xp"`
	Items       map[string]int `json:"items"`
}

// loadPlayQuestDependencies returns the dependency quest IDs for a play quest
// in insertion order. The caller must hold dbMu.
func loadPlayQuestDependencies(campaignID, questID string) ([]string, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT depends_on FROM play_quest_dependencies WHERE campaign_id=%s AND quest_id=%s ORDER BY id;", sq(campaignID), sq(questID)))
	if err != nil {
		return nil, err
	}
	var rows []struct {
		DependsOn string `json:"depends_on"`
	}
	if err := json.Unmarshal(out, &rows); err != nil {
		return nil, err
	}
	deps := make([]string, 0, len(rows))
	for _, row := range rows {
		deps = append(deps, row.DependsOn)
	}
	return deps, nil
}

// loadPlayQuest loads a single play quest by campaign and quest id. The
// caller must hold dbMu.
func loadPlayQuest(campaignID, questID string) (*playQuestResponse, bool, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT quest_id, title, state, rewards_xp, rewards_items FROM play_quests WHERE campaign_id=%s AND quest_id=%s LIMIT 1;", sq(campaignID), sq(questID)))
	if err != nil {
		return nil, false, err
	}
	var rows []struct {
		QuestID      string `json:"quest_id"`
		Title        string `json:"title"`
		State        string `json:"state"`
		RewardsXP    int    `json:"rewards_xp"`
		RewardsItems string `json:"rewards_items"`
	}
	if err := json.Unmarshal(out, &rows); err != nil {
		return nil, false, err
	}
	if len(rows) == 0 {
		return nil, false, nil
	}
	deps, err := loadPlayQuestDependencies(campaignID, questID)
	if err != nil {
		return nil, false, err
	}
	if deps == nil {
		deps = []string{}
	}
	rewards := questRewards{XP: rows[0].RewardsXP, Items: map[string]int{}}
	if rows[0].RewardsItems != "" && rows[0].RewardsItems != "null" {
		if err := json.Unmarshal([]byte(rows[0].RewardsItems), &rewards.Items); err != nil {
			return nil, false, err
		}
	}
	if rewards.Items == nil {
		rewards.Items = map[string]int{}
	}
	resp := &playQuestResponse{
		QuestID:   rows[0].QuestID,
		Title:     rows[0].Title,
		DependsOn: deps,
		State:     rows[0].State,
	}
	if questRewardsConfigured(rewards) {
		resp.Rewards = &rewards
	}
	return resp, true, nil
}

// playQuestDependenciesCompleted reports whether every dependency of the given
// quest has state 'completed'. The caller must hold dbMu.
func playQuestDependenciesCompleted(campaignID, questID string) (bool, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT COUNT(*) AS pending FROM play_quest_dependencies d JOIN play_quests q ON q.campaign_id=d.campaign_id AND q.quest_id=d.depends_on WHERE d.campaign_id=%s AND d.quest_id=%s AND q.state!='completed';", sq(campaignID), sq(questID)))
	if err != nil {
		return false, err
	}
	var rows []struct {
		Pending int `json:"pending"`
	}
	if err := json.Unmarshal(out, &rows); err != nil {
		return false, err
	}
	if len(rows) == 0 {
		return true, nil
	}
	return rows[0].Pending == 0, nil
}

// questRewardsConfigured reports whether a quest has non-default rewards.
func questRewardsConfigured(rewards questRewards) bool {
	return rewards.XP > 0 || len(rewards.Items) > 0
}

// createPlayQuestHandler lets the campaign DM create a quest with
// dependencies. Players receive 403. Duplicate quest IDs in the same campaign
// return 409. Invalid dependencies return 400.
func createPlayQuestHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")

	if _, ok := requireCampaignOwner(w, r, campaignID); !ok {
		return
	}

	var req playQuestCreateRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.QuestID == "" || req.Title == "" {
		writeError(w, http.StatusBadRequest, "invalid quest")
		return
	}

	seen := map[string]bool{}
	for _, dep := range req.DependsOn {
		if dep == "" {
			writeError(w, http.StatusBadRequest, "invalid dependency")
			return
		}
		if seen[dep] {
			writeError(w, http.StatusBadRequest, "invalid dependency")
			return
		}
		seen[dep] = true
		if dep == req.QuestID {
			writeError(w, http.StatusBadRequest, "invalid dependency")
			return
		}
	}

	dup, err := queryExists(fmt.Sprintf("SELECT 1 FROM play_quests WHERE campaign_id=%s AND quest_id=%s LIMIT 1;", sq(campaignID), sq(req.QuestID)))
	if err != nil {
		log.Printf("create play quest duplicate query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if dup {
		writeError(w, http.StatusConflict, "quest already exists")
		return
	}

	if len(req.DependsOn) > 0 {
		depQuoted := make([]string, len(req.DependsOn))
		for i, dep := range req.DependsOn {
			depQuoted[i] = sq(dep)
		}
		out, err := dbQuery(fmt.Sprintf("SELECT COUNT(*) AS cnt FROM play_quests WHERE campaign_id=%s AND quest_id IN (%s);", sq(campaignID), strings.Join(depQuoted, ", ")))
		if err != nil {
			log.Printf("create play quest dependency count query error: %v", err)
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
		var counts []struct {
			Cnt int `json:"cnt"`
		}
		if err := json.Unmarshal(out, &counts); err != nil {
			log.Printf("create play quest dependency count unmarshal error: %v", err)
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
		cnt := 0
		if len(counts) > 0 {
			cnt = counts[0].Cnt
		}
		if cnt != len(req.DependsOn) {
			writeError(w, http.StatusBadRequest, "invalid dependency")
			return
		}
	}

	var sb strings.Builder
	sb.WriteString("BEGIN;")
	sb.WriteString(fmt.Sprintf("INSERT INTO play_quests (campaign_id, quest_id, title, state) VALUES (%s, %s, %s, 'locked');", sq(campaignID), sq(req.QuestID), sq(req.Title)))
	for _, dep := range req.DependsOn {
		sb.WriteString(fmt.Sprintf("INSERT INTO play_quest_dependencies (campaign_id, quest_id, depends_on) VALUES (%s, %s, %s);", sq(campaignID), sq(req.QuestID), sq(dep)))
	}
	sb.WriteString("COMMIT;")
	if err := dbExec(sb.String()); err != nil {
		log.Printf("create play quest insert error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	respDependsOn := req.DependsOn
	if respDependsOn == nil {
		respDependsOn = []string{}
	}
	writeJSON(w, http.StatusCreated, playQuestResponse{
		QuestID:   req.QuestID,
		Title:     req.Title,
		DependsOn: respDependsOn,
		State:     "locked",
	})
}

// updatePlayQuestStateHandler lets the campaign DM change a quest's state.
// Players receive 403. Unknown quests return 404. Invalid state values return
// 400. Disallowed transitions return 409.
func updatePlayQuestStateHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")
	questID := r.PathValue("quest_id")

	if _, ok := requireCampaignOwner(w, r, campaignID); !ok {
		return
	}

	var req playQuestStateRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.State != "active" && req.State != "completed" {
		writeError(w, http.StatusBadRequest, "invalid state")
		return
	}

	quest, ok, err := loadPlayQuest(campaignID, questID)
	if err != nil {
		log.Printf("update play quest state load error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "quest not found")
		return
	}

	if req.State == "active" {
		if quest.State != "locked" {
			writeError(w, http.StatusConflict, "invalid transition")
			return
		}
		allCompleted, err := playQuestDependenciesCompleted(campaignID, questID)
		if err != nil {
			log.Printf("update play quest dependencies completed query error: %v", err)
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
		if !allCompleted {
			writeError(w, http.StatusConflict, "invalid transition")
			return
		}
	} else if req.State == "completed" {
		if quest.State != "active" {
			writeError(w, http.StatusConflict, "invalid transition")
			return
		}
	}

	if err := dbExec(fmt.Sprintf("UPDATE play_quests SET state=%s WHERE campaign_id=%s AND quest_id=%s;", sq(req.State), sq(campaignID), sq(questID))); err != nil {
		log.Printf("update play quest state update error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	updated, ok, err := loadPlayQuest(campaignID, questID)
	if err != nil {
		log.Printf("update play quest state reload error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "quest not found")
		return
	}
	writeJSON(w, http.StatusOK, updated)
}

// listPlayQuestsHandler returns all quests in a play campaign in creation
// order. It is available to the campaign owner and members.
func listPlayQuestsHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")

	if _, ok := requireCampaignOwnerOrMember(w, r, campaignID); !ok {
		return
	}

	out, err := dbQuery(fmt.Sprintf("SELECT quest_id, title, state, rewards_xp, rewards_items FROM play_quests WHERE campaign_id=%s ORDER BY id;", sq(campaignID)))
	if err != nil {
		log.Printf("list play quests query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	var rows []struct {
		QuestID      string `json:"quest_id"`
		Title        string `json:"title"`
		State        string `json:"state"`
		RewardsXP    int    `json:"rewards_xp"`
		RewardsItems string `json:"rewards_items"`
	}
	if err := json.Unmarshal(out, &rows); err != nil {
		log.Printf("list play quests unmarshal error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	quests := make([]playQuestResponse, 0, len(rows))
	for _, row := range rows {
		deps, err := loadPlayQuestDependencies(campaignID, row.QuestID)
		if err != nil {
			log.Printf("list play quests dependencies query error: %v", err)
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
		if deps == nil {
			deps = []string{}
		}
		rewards := questRewards{XP: row.RewardsXP, Items: map[string]int{}}
		if row.RewardsItems != "" && row.RewardsItems != "null" {
			if err := json.Unmarshal([]byte(row.RewardsItems), &rewards.Items); err != nil {
				log.Printf("list play quests rewards unmarshal error: %v", err)
				writeError(w, http.StatusInternalServerError, "internal error")
				return
			}
		}
		if rewards.Items == nil {
			rewards.Items = map[string]int{}
		}
		q := playQuestResponse{
			QuestID:   row.QuestID,
			Title:     row.Title,
			DependsOn: deps,
			State:     row.State,
		}
		if questRewardsConfigured(rewards) {
			q.Rewards = &rewards
		}
		quests = append(quests, q)
	}
	if quests == nil {
		quests = []playQuestResponse{}
	}

	writeJSON(w, http.StatusOK, playQuestListResponse{Quests: quests})
}

// configureQuestRewardsHandler lets the campaign DM configure rewards for a
// quest. Players receive 403. Unknown quests return 404. The quest must be in
// state 'locked' or 'active'; completed quests reject configuration with 409.
// Invalid reward bodies return 400.
func configureQuestRewardsHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")
	questID := r.PathValue("quest_id")

	if _, ok := requireCampaignOwner(w, r, campaignID); !ok {
		return
	}

	var req questRewardsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Items == nil || req.XP < 0 {
		writeError(w, http.StatusBadRequest, "invalid rewards")
		return
	}
	for itemID, qty := range req.Items {
		if !validInventoryItemIDs[itemID] || qty <= 0 {
			writeError(w, http.StatusBadRequest, "invalid rewards")
			return
		}
	}

	quest, ok, err := loadPlayQuest(campaignID, questID)
	if err != nil {
		log.Printf("configure quest rewards load error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "quest not found")
		return
	}
	if quest.State != "locked" && quest.State != "active" {
		writeError(w, http.StatusConflict, "quest cannot be configured")
		return
	}

	itemsJSON, err := json.Marshal(req.Items)
	if err != nil {
		log.Printf("configure quest rewards items marshal error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	if err := dbExec(fmt.Sprintf("UPDATE play_quests SET rewards_xp=%d, rewards_items=%s WHERE campaign_id=%s AND quest_id=%s;",
		req.XP, sq(string(itemsJSON)), sq(campaignID), sq(questID))); err != nil {
		log.Printf("configure quest rewards update error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	updated, ok, err := loadPlayQuest(campaignID, questID)
	if err != nil {
		log.Printf("configure quest rewards reload error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "quest not found")
		return
	}
	writeJSON(w, http.StatusOK, updated)
}

// awardQuestRewardsHandler lets the campaign DM award configured quest rewards
// once to every campaign member. Players receive 403. Unknown quests return
// 404. The quest must be completed and configured, and rewards must not have
// been awarded before; otherwise the request returns 409.
func awardQuestRewardsHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")
	questID := r.PathValue("quest_id")

	if _, ok := requireCampaignOwner(w, r, campaignID); !ok {
		return
	}

	quest, ok, err := loadPlayQuest(campaignID, questID)
	if err != nil {
		log.Printf("award quest rewards load error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "quest not found")
		return
	}
	if quest.State != "completed" {
		writeError(w, http.StatusConflict, "quest not completed")
		return
	}
	if quest.Rewards == nil {
		writeError(w, http.StatusConflict, "rewards not configured")
		return
	}

	alreadyAwarded, err := queryExists(fmt.Sprintf("SELECT 1 FROM play_quests WHERE campaign_id=%s AND quest_id=%s AND rewards_awarded=1 LIMIT 1;", sq(campaignID), sq(questID)))
	if err != nil {
		log.Printf("award quest rewards awarded check error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if alreadyAwarded {
		writeError(w, http.StatusConflict, "rewards already awarded")
		return
	}

	members, err := queryPlayCampaignMembers(campaignID)
	if err != nil {
		log.Printf("award quest rewards members query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	itemsJSON, err := json.Marshal(quest.Rewards.Items)
	if err != nil {
		log.Printf("award quest rewards items marshal error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	var sb strings.Builder
	sb.WriteString("BEGIN;")
	for _, member := range members {
		sb.WriteString(fmt.Sprintf("INSERT INTO play_quest_reward_grants (campaign_id, quest_id, character_id, xp, items) VALUES (%s, %s, %s, %d, %s) ON CONFLICT(campaign_id, quest_id, character_id) DO UPDATE SET xp=excluded.xp, items=excluded.items;",
			sq(campaignID), sq(questID), sq(member.CharacterID), quest.Rewards.XP, sq(string(itemsJSON))))
		for itemID, qty := range quest.Rewards.Items {
			sb.WriteString(fmt.Sprintf(
				"INSERT INTO character_inventory_items (campaign_id, character_id, item_id, quantity) VALUES (%s, %s, %s, %d) ON CONFLICT(campaign_id, character_id, item_id) DO UPDATE SET quantity = quantity + excluded.quantity;",
				sq(campaignID), sq(member.CharacterID), sq(itemID), qty))
		}
	}
	sb.WriteString(fmt.Sprintf("UPDATE play_quests SET rewards_awarded=1 WHERE campaign_id=%s AND quest_id=%s;", sq(campaignID), sq(questID)))
	sb.WriteString("COMMIT;")
	if err := dbExec(sb.String()); err != nil {
		log.Printf("award quest rewards exec error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	writeJSON(w, http.StatusCreated, questAwardResponse{
		QuestID: questID,
		Awarded: true,
		XP:      quest.Rewards.XP,
		Items:   quest.Rewards.Items,
	})
}

// getCharacterQuestRewardsHandler returns cumulative quest reward grants for a
// campaign character. It is available to authenticated campaign members.
func getCharacterQuestRewardsHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	campaignID := r.PathValue("id")
	characterID := r.PathValue("character_id")

	if _, ok := requireCampaignOwnerOrMember(w, r, campaignID); !ok {
		return
	}

	_, ok, err := queryPlayCampaignMember(campaignID, characterID)
	if err != nil {
		log.Printf("get character quest rewards member query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "character not found")
		return
	}

	out, err := dbQuery(fmt.Sprintf("SELECT xp, items FROM play_quest_reward_grants WHERE campaign_id=%s AND character_id=%s;", sq(campaignID), sq(characterID)))
	if err != nil {
		log.Printf("get character quest rewards query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	var rows []struct {
		XP    int    `json:"xp"`
		Items string `json:"items"`
	}
	if err := json.Unmarshal(out, &rows); err != nil {
		log.Printf("get character quest rewards unmarshal error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	totalXP := 0
	totalItems := map[string]int{}
	for _, row := range rows {
		totalXP += row.XP
		if row.Items != "" && row.Items != "null" {
			var items map[string]int
			if err := json.Unmarshal([]byte(row.Items), &items); err != nil {
				log.Printf("get character quest rewards items unmarshal error: %v", err)
				writeError(w, http.StatusInternalServerError, "internal error")
				return
			}
			for itemID, qty := range items {
				totalItems[itemID] += qty
			}
		}
	}

	writeJSON(w, http.StatusOK, characterQuestRewardsResponse{
		CharacterID: characterID,
		XP:          totalXP,
		Items:       totalItems,
	})
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T002352Z_pi_kimi-k2p7-code_python-django/dndsite/views/play_quests.py
"""Live-play campaign quest dependency views."""

import json
import sqlite3

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from .common import (
    require_play_campaign_dm_owner,
    require_play_campaign_member_or_owner,
)
from .inventory import VALID_ITEM_IDS
from ..http import bad_request, conflict, not_found, parse_json, require_method
from ..db import (
    _configure_quest_rewards,
    _grant_quest_rewards_to_member,
    _is_quest_awarded,
    _record_quest_award,
    db_conn,
)


VALID_STATES = {"active", "completed"}


def _quest_record(row):
    record = {
        "quest_id": row["quest_id"],
        "title": row["title"],
        "depends_on": json.loads(row["depends_on_json"]),
        "state": row["state"],
    }
    try:
        rewards_json = row["rewards_json"]
    except (KeyError, IndexError):
        rewards_json = None
    if rewards_json and rewards_json != '{}':
        try:
            record["rewards"] = json.loads(rewards_json)
        except (json.JSONDecodeError, TypeError):
            pass
    return record


@csrf_exempt
def play_quests(request, id):
    bad = require_method(request, "GET", "POST")
    if bad:
        return bad
    if request.method == "POST":
        return _create_play_quest(request, id)
    return _list_play_quests(request, id)


def _create_play_quest(request, id):
    actor, campaign, err = require_play_campaign_dm_owner(request, id)
    if err is not None:
        return err

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")

    try:
        quest_id = body["quest_id"]
        title = body["title"]
        depends_on = body["depends_on"]
        if not isinstance(quest_id, str) or not quest_id:
            raise ValueError
        if not isinstance(title, str) or not title:
            raise ValueError
        if not isinstance(depends_on, list):
            raise ValueError
        if any(not isinstance(d, str) or not d for d in depends_on):
            raise ValueError
        if len(depends_on) != len(set(depends_on)):
            raise ValueError
        if quest_id in depends_on:
            raise ValueError
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    with db_conn() as conn:
        if depends_on:
            placeholders = ",".join("?" * len(depends_on))
            rows = conn.execute(
                f"SELECT quest_id FROM play_quests "
                f"WHERE campaign_id = ? AND quest_id IN ({placeholders})",
                (id, *depends_on),
            ).fetchall()
            found = {row["quest_id"] for row in rows}
            if found != set(depends_on):
                return bad_request("invalid request")

        try:
            conn.execute(
                "INSERT INTO play_quests (campaign_id, quest_id, title, depends_on_json, state) "
                "VALUES (?, ?, ?, ?, ?)",
                (id, quest_id, title, json.dumps(depends_on), "locked"),
            )
        except sqlite3.IntegrityError:
            return conflict("quest already exists")

    return JsonResponse(
        {
            "quest_id": quest_id,
            "title": title,
            "depends_on": depends_on,
            "state": "locked",
        },
        status=201,
    )


@csrf_exempt
def update_play_quest_state(request, id, quest_id):
    bad = require_method(request, "PUT")
    if bad:
        return bad

    actor, campaign, err = require_play_campaign_dm_owner(request, id)
    if err is not None:
        return err

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")

    try:
        new_state = body["state"]
        if new_state not in VALID_STATES:
            raise ValueError
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    with db_conn() as conn:
        row = conn.execute(
            "SELECT id, title, depends_on_json, state, rewards_json FROM play_quests "
            "WHERE campaign_id = ? AND quest_id = ?",
            (id, quest_id),
        ).fetchone()
        if row is None:
            return not_found("quest not found")

        current_state = row["state"]
        depends_on = json.loads(row["depends_on_json"])
        allowed = False

        if current_state == "locked" and new_state == "active":
            if depends_on:
                placeholders = ",".join("?" * len(depends_on))
                dep_rows = conn.execute(
                    f"SELECT quest_id, state FROM play_quests "
                    f"WHERE campaign_id = ? AND quest_id IN ({placeholders})",
                    (id, *depends_on),
                ).fetchall()
                dep_states = {r["quest_id"]: r["state"] for r in dep_rows}
                if len(dep_states) != len(depends_on):
                    return conflict("invalid transition")
                if all(dep_states[d] == "completed" for d in depends_on):
                    allowed = True
            else:
                allowed = True
        elif current_state == "active" and new_state == "completed":
            allowed = True

        if not allowed:
            return conflict("invalid transition")

        conn.execute(
            "UPDATE play_quests SET state = ? WHERE campaign_id = ? AND quest_id = ?",
            (new_state, id, quest_id),
        )

        record = {
            "quest_id": quest_id,
            "title": row["title"],
            "depends_on": depends_on,
            "state": new_state,
        }
        rewards_json = row["rewards_json"]
        if rewards_json and rewards_json != '{}':
            try:
                record["rewards"] = json.loads(rewards_json)
            except (json.JSONDecodeError, TypeError):
                pass

        return JsonResponse(record)


def _list_play_quests(request, id):
    actor, campaign, err = require_play_campaign_member_or_owner(request, id)
    if err is not None:
        return err

    with db_conn() as conn:
        rows = conn.execute(
            "SELECT quest_id, title, depends_on_json, state, rewards_json FROM play_quests "
            "WHERE campaign_id = ? ORDER BY id",
            (id,),
        ).fetchall()

    return JsonResponse({"quests": [_quest_record(row) for row in rows]})


@csrf_exempt
def configure_quest_rewards(request, id, quest_id):
    bad = require_method(request, "PUT")
    if bad:
        return bad

    actor, campaign, err = require_play_campaign_dm_owner(request, id)
    if err is not None:
        return err

    body = parse_json(request)
    if body is None:
        return bad_request("invalid request")

    try:
        xp = body["xp"]
        items = body["items"]
        if not isinstance(xp, int) or isinstance(xp, bool) or xp < 0:
            raise ValueError
        if not isinstance(items, dict):
            raise ValueError
        if any(not isinstance(k, str) or k not in VALID_ITEM_IDS for k in items.keys()):
            raise ValueError
        if any(not isinstance(v, int) or isinstance(v, bool) or v <= 0 for v in items.values()):
            raise ValueError
    except (KeyError, TypeError, ValueError):
        return bad_request("invalid request")

    rewards = {"xp": xp, "items": items}

    with db_conn() as conn:
        row = conn.execute(
            "SELECT id, title, depends_on_json, state, rewards_json FROM play_quests "
            "WHERE campaign_id = ? AND quest_id = ?",
            (id, quest_id),
        ).fetchone()
        if row is None:
            return not_found("quest not found")

        if row["state"] == "completed":
            return conflict("quest already completed")

        _configure_quest_rewards(conn, id, quest_id, rewards)

    return JsonResponse(
        {
            "quest_id": quest_id,
            "title": row["title"],
            "depends_on": json.loads(row["depends_on_json"]),
            "state": row["state"],
            "rewards": rewards,
        }
    )


@csrf_exempt
def award_quest_rewards(request, id, quest_id):
    bad = require_method(request, "POST")
    if bad:
        return bad

    actor, campaign, err = require_play_campaign_dm_owner(request, id)
    if err is not None:
        return err

    with db_conn() as conn:
        row = conn.execute(
            "SELECT id, title, depends_on_json, state, rewards_json FROM play_quests "
            "WHERE campaign_id = ? AND quest_id = ?",
            (id, quest_id),
        ).fetchone()
        if row is None:
            return not_found("quest not found")

        if row["state"] != "completed":
            return conflict("quest not completed")

        rewards_json = row["rewards_json"]
        if not rewards_json or rewards_json == '{}':
            return conflict("rewards not configured")

        try:
            rewards = json.loads(rewards_json)
        except (json.JSONDecodeError, TypeError):
            return conflict("rewards not configured")

        xp = rewards.get("xp", 0)
        items = rewards.get("items", {})

        if _is_quest_awarded(conn, id, quest_id):
            return conflict("rewards already awarded")

        members = conn.execute(
            "SELECT character_id FROM play_campaign_members WHERE campaign_id = ?",
            (id,),
        ).fetchall()

        for member in members:
            _grant_quest_rewards_to_member(conn, id, member["character_id"], xp, items)

        _record_quest_award(conn, id, quest_id)

    return JsonResponse(
        {
            "quest_id": quest_id,
            "awarded": True,
            "xp": xp,
            "items": items,
        },
        status=201,
    )


@csrf_exempt
def get_character_quest_rewards(request, id, character_id):
    bad = require_method(request, "GET")
    if bad:
        return bad

    actor, campaign, err = require_play_campaign_member_or_owner(request, id)
    if err is not None:
        return err

    with db_conn() as conn:
        member = conn.execute(
            "SELECT character_id, quest_rewards_xp, quest_rewards_items_json "
            "FROM play_campaign_members WHERE campaign_id = ? AND character_id = ?",
            (id, character_id),
        ).fetchone()
        if member is None:
            return not_found("character not found")

    return JsonResponse(
        {
            "character_id": member["character_id"],
            "xp": member["quest_rewards_xp"],
            "items": json.loads(member["quest_rewards_items_json"]),
        }
    )
